src/posix/platform/openthread-posix-config.h

```diff
@@ -441,4 +441,25 @@
 #ifdef OPENTHREAD_POSIX_CONFIG_RCP_TIME_SYNC_INTERVAL
 #error "OPENTHREAD_POSIX_CONFIG_RCP_TIME_SYNC_INTERVAL was replaced by OPENTHREAD_SPINEL_CONFIG_RCP_TIME_SYNC_INTERVAL"
 #endif
+
+/**
+ * @def OPENTHREAD_POSIX_CONFIG_SETUP_INFRA_NETIF_AT_STARTUP_ENABLE
+ *
+ * Defines whether to setup the features related to infrastructure network interface at start up. If true, it requires
+ * the infrastructure network interface to be available when the process starts. If false, the platform is responsible
+ * to call `otSysInfraNetifInit` and `otSysInfraNetifSetUp` when the infrastructure network interface is ready.
+ *
+ */
```

Comment:
Thanks for pointing this out.

My thought is that when the platform initialization is deferred, the `otSysInfraNetif*` caller should be responsible for NOT directly/indirectly calling `otPlat*` APIs of BR/TREL/BBR before calling `otSysInfraNetifInit`. So the only risk of calling `otPlat*` APIs of uninitialized modules is during the `otInstanceInitSingle` call. If `otInstanceInitSingle` doesn't involve any `otPlat*` call of BR/TREL/BBR, can it be regarded as safe? 

Besides, I also have a question about re-initializing the infra netif: Is it valid if I call the APIs to switch to another infra link (e.g. from `eth0` to `wlan0`)? Does OT have any assumption about constant infra link?
```
    otSysInfraNetifTearDown();
    otSysInfraNetifDeinit();
    otSysInfraNetifInit(&platformConfig, icmpSocket);
    otSysInfraNetifSetUp(GetOtInstance());
```

Reply:
> Does OT have any assumption about constant infra link?

We have `otPlatInfraIfStateChanged()` which signals state changes of infra-if. OT stack itself does not require or assume that infra-if should be up and ready on its start. It waits for this to be signaled through this platform API.

> Is it valid if I call the APIs to switch to another infra link (e.g. from eth0 to wlan0)?

We have the `otBorderRoutingInit()` API (it is not a platform API, but it is a public API) which sets the infra-if for `BorderRouting` function. It can be called again to change the infra-if.
https://github.com/openthread/openthread/blob/6bfe59dabf1cf4c2b02d950418a218a6a030d6c2/include/openthread/border_routing.h#L135-L137

------

> If otInstanceInitSingle doesn't involve any otPlat* call of BR/TREL/BBR, can it be regarded as safe?

Not sure this would work and it adds a hard requirement on OT core to not call certain `otPlat` APIs at  certain times  But then after `InitSingle()`, we start the main run loop which allows OT code stack to run. OT core can then call any `otPlat` API it wants. We have no way of having a "otSysInfraNetif* caller" stopping this.

---

Can you maybe share more details about the background of this change? What is the problem we are trying to solve with this? 

My overall suggestion/thoughts on this:
- The existing model where we initialize all platform modules before initing `otInstance` and starting main loop is a good and clean approach. So I highly suggest keeping it as is. 
- If we think certain platforms may need more time to be ready, we should consider adding mechanism to handle this (like the infra-if callback which inform its state change and tell when its ready).

Reply:
>Can you maybe share more details about the background of this change? What is the problem we are trying to solve with this?

Sure. On Android, we may allow the TBR to use either Ethernet or WiFi as the infra link. So it's possible that the TBR switches to another infra link, e.g. when the user unplugs the Ethernet cable and connects to WiFi. Also, the infra link may not be available at the time the OT starts. That's why we want to defer the initialization of the infra link and make it able to switch infra link during runtime.

BTW, there's already a [mechanism](https://github.com/openthread/ot-br-posix/blob/main/src/utils/infra_link_selector.cpp) to restart otbr-agent when switching to another infra link but it requires the whole process to restart so it may not be preferred if any on-the-fly approach is available. 

> We have the otBorderRoutingInit() API (it is not a platform API, but it is a public API) which sets the infra-if for BorderRouting function. It can be called again to change the infra-if.

I noticed this one when investigating the code but I doubt if it works as expected. From my understanding the `ot::Posix::InfraNetif` component needs to be updated (set netif index, create ICMPv6 socket, etc.) accordingly when switching to another infra netif. However, in the call path of `otBorderRoutingInit` I didn't see it updating `ot::Posix::InfraNetif`.

Reply:
Thanks.  I think it may be better to come up with a solution (new APIs or mechanisms) related to this and defering the platform initialization may not a good way. We want all platform modules to always be initialized first before we start OT stack but can add new mechanisms to ask them to update or change. 

Here are two ways I can suggest:

- I expect a software component on the device or platform to control and determine which netif index to use.
- This component is the one that first calls `otBorderRoutingInit()` and later inform of the the index changes.
- We can either add a mechanism (e.g., a new API or callback) for this software component to inform the platform layer as well.
- Or we can add a new OT platform API (say `otPlatInfraIfInit()`) for the OT stack to inform the platform layer that its public API `otBorderRoutingInit()` is called and to inform the platform of the infra-if index it is instructed to use for border routing.
- The platform can then make any changes it needs to do when informed (e.g., create a new ICMP socket, etc.).

----

- We can extend the platform implementation to be aware and support multiple netif indexes and be ready for any to be used. 
- Basically making sure `otPlatInfraIf{}` APIs can work with different `aInfraIfIndex` inputs.
- Calling `otBorderRoutingInit()` would sets the infra-if index to use on OT stack. 
- OT stack will then use `otPlatInfraIf` APIs which always get `aInfraIfIndex` as input.
- If the infra-if index is changed, then OT stack will call `otPlatInfraIf` with new index and platform should know and support this index already.

Reply:
> I expect a software component on the device or platform to control and determine which netif index to use.

Yes. Let's call that component `ThreadNetworkService`. It will determine which infra link to use and notify OT stack by OT API. Also, due to SELinux policies, the ICMPv6 socket on infra link has to be created by `ThreadNetworkService` (which is in a privileged process) and `ThreadNetworkService` will pass the socket to OT stack via `otSysInfraNetifInit` API.

In both option 1 and option 2, I'm not sure when we should call `ot::Posix::InfraNetif::Init(const char *ainfraIfName)`? According to your suggestion this should be called before the `otInstance` is created. However, we don't have a valid/meaningful infra netif at that moment. Should we allow using an invalid infra link as the input? It apparently makes the `ot::Posix::InfraNetif::Init` less useful.

Or probably we should move the work of `ot::Posix::InfraNetif::Init` to a platform API which is triggered by `otBorderRoutingInit` (and similarly for BBR and TREL)?  In this way we can get rid of having order dependencies on `platform(BR/TREL/BBR)Init` functions. It requires quite some work and may be risky though. Also here's anoter problem about how to pass the socket to OT stack: Socket FD is platform dependent so it may not be a good parameter for `otBorderRoutingInit`.

Reply:
Here are my suggestions:
- Platform modules should always be initialized before initializing the `otInstance`. Delaying initialization could be risky and error-prone.
- However, platform initialization itself does not need to handle everything.
- We want to ensure that all platforms are initialized so they can set their local variables, state, etc., and be ready to accept otPlat API calls.
- Platform code can wait for other calls to take more actions.

> Or probably we should move the work of `ot::Posix::InfraNetif::Init` to a platform API which is triggered by otBorderRoutingInit (and similarly for BBR and TREL)? 

Yes, that is the approach that makes the most sense to me and that I would recommend. It will require new code to support this, but it is definitely less risky than leaving the platform uninitialized and deferring its initialization. This model can also support later changes to the infrastructure-if index or similar situations.

Reply:
Thanks for the suggestions! 

- I've narrowed the scope to `ot::Posix::InfraNetif` and decoupled the `ot::Posix::InfraNetif::Init()` from specifying the infra link. In this way we're not changing the order of `platformInit()`, `otInstanceInitSingle()` and `platformSetUp()`. 
- I also allowed the `RoutingManager` to init with 0 as the interface index if the infra link is not explicitly specified. This should be fine because 0 can be regarded as an interface that never goes up and running.